File: test/language/parameterized_types/python/ArrayElementParamTest.py
import unittest
import zserio

from testutils import getZserioApi

class ArrayElementParamTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "parameterized_types.zs").array_element_param

    def testWrite(self):
        database = self._createDatabase()
        writer = zserio.BitStreamWriter()
        database.initialize_offsets(writer.bitposition)
        database.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        self._checkDatabaseInStream(reader, database)

        reader.bitposition = 0
        readDatabase = self.api.Database.from_reader(reader)
        self.assertEqual(database, readDatabase)

    def _createDatabase(self):
        blockHeaderList = []
        blockList = []
        for i in range(self.NUM_BLOCKS):
            numItems = i + 1
            blockHeader = self.api.BlockHeader(numItems, 0)
            blockHeaderList.append(blockHeader)
            itemList = []
            for j in range(numItems):
                itemList.append(j * 2)

            blockList.append(self.api.Block(blockHeader, itemList))

        return self.api.Database(self.NUM_BLOCKS, blockHeaderList, blockList)

    def _checkDatabaseInStream(self, reader, database):
        numBlocks = database.num_blocks
        self.assertEqual(numBlocks, reader.read_bits(16))

        headers = database.headers
        expectedOffset = self.FIRST_BYTE_OFFSET
        for i in range(numBlocks):
            numItems = reader.read_bits(16)
            self.assertEqual(headers[i].num_items, numItems)
            self.assertEqual(expectedOffset, reader.read_bits(32))
            expectedOffset += 8 * numItems

        blocks = database.blocks
        for i in range(numBlocks):
            numItems = headers[i].num_items
            items = blocks[i].items
            for j in range(numItems):
                self.assertEqual(items[j], reader.read_signed_bits(64))

    NUM_BLOCKS = 3
    FIRST_BYTE_OFFSET = 2 + NUM_BLOCKS * (2 + 4)

File: test/language/parameterized_types/cpp/ArrayElementParamTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "parameterized_types/array_element_param/Database.h"

namespace parameterized_types
{
namespace array_element_param
{

class ArrayElementParamTest : public ::testing::Test
{
protected:
    void fillDatabase(Database& database)
    {
        database.setNumBlocks(NUM_BLOCKS);
        auto& headers = database.getHeaders();
        for (uint16_t i = 0; i < NUM_BLOCKS; ++i)
        {
            BlockHeader blockHeader;
            blockHeader.setNumItems(i + 1);
            blockHeader.setOffset(0);
            headers.push_back(blockHeader);
        }

        auto& blocks = database.getBlocks();
        for (auto it = headers.begin(); it != headers.end(); ++it)
        {
            Block block;
            auto& items = block.getItems();
            const uint16_t numItems = it->getNumItems();
            for (uint16_t j = 0; j < numItems; ++j)
                items.push_back(j * 2);
            blocks.push_back(block);
        }

        database.initializeChildren();
        database.initializeOffsets();
    }

    void checkDatabaseInBitStream(zserio::BitStreamReader& reader, const Database& database)
    {
        const uint16_t numBlocks = database.getNumBlocks();

        ASSERT_EQ(numBlocks, reader.readBits(16));

        const auto& headers = database.getHeaders();
        uint32_t expectedOffset = FIRST_BYTE_OFFSET;
        for (uint16_t i = 0; i < numBlocks; ++i)
        {
            const uint16_t numItems = static_cast<uint16_t>(reader.readBits(16));
            ASSERT_EQ(headers.at(i).getNumItems(), numItems);
            ASSERT_EQ(expectedOffset, reader.readBits(32));
            expectedOffset += 8 * numItems;
        }

        const auto& blocks = database.getBlocks();
        for (uint16_t i = 0; i < numBlocks; ++i)
        {
            const uint16_t numItems = headers.at(i).getNumItems();
            const auto& items = blocks.at(i).getItems();
            for (uint16_t j = 0; j < numItems; ++j)
                ASSERT_EQ(items.at(j), reader.readBits64(64));
        }
    }

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);

private:
    static const uint16_t NUM_BLOCKS;
    static const uint32_t FIRST_BYTE_OFFSET;
};

const uint16_t ArrayElementParamTest::NUM_BLOCKS = 3;
const uint32_t ArrayElementParamTest::FIRST_BYTE_OFFSET =
        2 + ArrayElementParamTest::NUM_BLOCKS * (2 + 4);

TEST_F(ArrayElementParamTest, fieldConstructor)
{
    Database database;
    fillDatabase(database);
    // initialize because Block::operator== touches header parameter (see last assert)
    database.initializeChildren();

    auto headers = database.getHeaders();
    auto blocks = database.getBlocks();

    void* headersPtr = headers.data();
    void* blocksPtr = blocks.data();

    // headers are moved, blocks copied
    Database newDatabase(database.getNumBlocks(), std::move(headers), blocks);
    ASSERT_EQ(headersPtr, newDatabase.getHeaders().data());
    ASSERT_NE(blocksPtr, newDatabase.getBlocks().data());
    ASSERT_EQ(blocks, newDatabase.getBlocks());
}

TEST_F(ArrayElementParamTest, moveConstructor)
{
    Database database;
    fillDatabase(database);

    void* headersPtr = database.getHeaders().data();
    void* blocksPtr = database.getBlocks().data();

    Database movedDatabase(std::move(database));
    ASSERT_EQ(headersPtr, movedDatabase.getHeaders().data());
    ASSERT_EQ(blocksPtr, movedDatabase.getBlocks().data());
}

TEST_F(ArrayElementParamTest, moveAssignmentOperator)
{
    Database database;
    fillDatabase(database);

    void* headersPtr = database.getHeaders().data();
    void* blocksPtr = database.getBlocks().data();

    Database movedDatabase;
    movedDatabase = std::move(database);
    ASSERT_EQ(headersPtr, movedDatabase.getHeaders().data());
    ASSERT_EQ(blocksPtr, movedDatabase.getBlocks().data());
}

TEST_F(ArrayElementParamTest, write)
{
    Database database;
    fillDatabase(database);

    zserio::BitStreamWriter writer(bitBuffer);
    database.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    checkDatabaseInBitStream(reader, database);
    reader.setBitPosition(0);

    Database readDatabase(reader);
    ASSERT_EQ(database, readDatabase);
}

} // namespace array_element_param
} // namespace parameterized_types
